contributors/design-proposals/pod-priority-api.md

```diff
@@ -0,0 +1,237 @@
+string and all of the valid priority classes are defined by a system wide
+mapping that maps each string to an integer. The PriorityClassName specified in
+a pod spec must be found in this map or the pod creation request will be
+rejected. If PriorityClassName is empty, it will resolve to the default
+priority (See below for more info on name resolution). Once the 
+PriorityClassName is resolved to an integer, it is placed in "Priority" field of
+PodSpec.
+
+
+```
+type PodSpec struct {
+  ...
+  PriorityClassName string
+  Priority          int32  // Populated by Admission Controller. Users are not allowed to set it directly.
+}
+```
+
+### Priority Classes 
+
+A Kubernetes cluster has one predefined (built-in) priority class which is 
+reserved for the system pods and its value is the largest 32-bit
+integer. The cluster may have many more user defined priority classes for
+various use cases. The following list is an example of how the priorities and
+their values may look like.
+We decided to ship Kubernetes with no predefined priority classes, except
+for the "system" priority which will be built in Kubernetes. The main goal of
+of having no built-in priority classes is to avoid creating defacto standard
+names which may be hard to change in the future.
+
+```
+system  2147483647 (int_max)
+tier1   4000
+tier2   2000
+tier3   1000
+```
+
+The following shows a list of example workloads in a Kubernetes cluster in decreasing order of priority:
+
+* Kubernetes system daemons (per-node like fluentd, and cluster-level like 
+  Heapster)
+* Critical user infrastructure (e.g. storage servers, monitoring system like
+  Prometheus, etc.)
+* Components that are in the user-facing request serving path and must be able 
+  to scale up arbitrarily in response to load spikes (web servers, middleware,
+  etc.)
+* Important interruptible workloads that need strong guarantee of
+  schedulability and of not being interrupted
+* Less important interruptible workloads that need a less strong guarantee of
+  schedulability and of not being interrupted
+* Best effort / opportunistic
+
+### Resolving priority class names 
+
+User requests sent to Kubernetes may have "PriorityClassName" in their PodSpec.
+Admission controller resolves a PriorityCalssName to its corresponding number
+and populates the "Priority" field of the pod spec. The rest of Kubernetes
+components look at the "Priority" field of pod status and work with the integer
+value. In other words, "PriorityClassName" will be ignored by the rest of the
+system.
+
+We are going to add a new API object called PriorityClass. The priority class
+defines the mapping between the priority name and its value. It can have an
+optional description. The description is an arbitrary string and is provided
+only as a guideline for users.
+
+If a pod does not specify any "PriorityClassName", the system resolves it to 
+zero in the first version, but we will add support to annotate one of the
```

Comment:
> "undefined" means no PriorityClass for zero; and real priority of "zero" is unknown, e.g. if all priority are negative, zero is highest priority :).

I am not sure what you mean by "real priority of zero is unknown". None of the components (other than a particular admission controller) should care about PriorityClass and should work with the integer value of priority. Zero is a valid integer and if users decide to run all their pods with negative priority, it will be totally fine. Of course, zero will be the highest priority in that case, but why is that a concern?